src/vulkan/device.cc

```diff
@@ -403,121 +402,134 @@ Result Device::Initialize(
-        storage16_ptrs =
-            static_cast<VkPhysicalDevice16BitStorageFeaturesKHR*>(ptr);
-      } else if (
-          s->sType ==
-          VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SUBGROUP_SIZE_CONTROL_FEATURES_EXT) {  // NOLINT(whitespace/line_length)
-        subgroup_size_control_features =
-            static_cast<VkPhysicalDeviceSubgroupSizeControlFeaturesEXT*>(ptr);
-      }
-      ptr = s->pNext;
+  // Search for additional features in case they are found in pNext field of
+  // available_features2.
+  VkPhysicalDeviceVariablePointerFeaturesKHR* var_ptrs = nullptr;
+  VkPhysicalDeviceFloat16Int8FeaturesKHR* float16_ptrs = nullptr;
+  VkPhysicalDevice8BitStorageFeaturesKHR* storage8_ptrs = nullptr;
+  VkPhysicalDevice16BitStorageFeaturesKHR* storage16_ptrs = nullptr;
+  VkPhysicalDeviceVulkan11Features* vulkan11_ptrs = nullptr;
+  VkPhysicalDeviceVulkan12Features* vulkan12_ptrs = nullptr;
+  VkPhysicalDeviceSubgroupSizeControlFeaturesEXT*
+      subgroup_size_control_features = nullptr;
+  void* ptr = available_features2.pNext;
+  while (ptr != nullptr) {
+    BaseOutStructure* s = static_cast<BaseOutStructure*>(ptr);
+    if (s->sType ==
+        VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VARIABLE_POINTER_FEATURES_KHR) {
+      var_ptrs = static_cast<VkPhysicalDeviceVariablePointerFeaturesKHR*>(ptr);
+    } else if (s->sType ==
+               VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_FLOAT16_INT8_FEATURES_KHR) {
+      float16_ptrs = static_cast<VkPhysicalDeviceFloat16Int8FeaturesKHR*>(ptr);
+    } else if (s->sType ==
+               VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_8BIT_STORAGE_FEATURES_KHR) {
+      storage8_ptrs = static_cast<VkPhysicalDevice8BitStorageFeaturesKHR*>(ptr);
+    } else if (s->sType ==
+               VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_16BIT_STORAGE_FEATURES_KHR) {
+      storage16_ptrs =
+          static_cast<VkPhysicalDevice16BitStorageFeaturesKHR*>(ptr);
+    } else if (
+        s->sType ==
+        VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_SUBGROUP_SIZE_CONTROL_FEATURES_EXT) {  // NOLINT(whitespace/line_length)
+      subgroup_size_control_features =
+          static_cast<VkPhysicalDeviceSubgroupSizeControlFeaturesEXT*>(ptr);
+    } else if (s->sType ==
+               VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VULKAN_1_1_FEATURES) {
+      vulkan11_ptrs = static_cast<VkPhysicalDeviceVulkan11Features*>(ptr);
+    } else if (s->sType ==
+               VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_VULKAN_1_2_FEATURES) {
+      vulkan12_ptrs = static_cast<VkPhysicalDeviceVulkan12Features*>(ptr);
     }
+    ptr = s->pNext;
+  }
 
-    std::vector<std::string> required_features1;
-    for (const auto& feature : required_features) {
-      // No dot means this is a features1 feature.
-      if (feature.find_first_of('.') == std::string::npos) {
-        required_features1.push_back(feature);
-        continue;
-      }
+  // Compare the available additional (non-core) features against the
+  // requirements.
```

Comment:
Could you add another comment that is similar to your PR description. E.g.

Vulkan 1.2 added support for defining non-core physical device features using VkPhysicalDeviceVulkan11Features and VkPhysicalDeviceVulkan12Features structures. If |vulkan11_ptrs| and/or |vulkan12_ptrs| are null, we must check for features using the old approach (by checking across various feature structs); otherwise, we can check features via the new structs.

Reply:
done